File: backend/models/student.model.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/db.config');
const Institute = require('./institute.model');
const Department = require('./department.model');
const Section = require('./section.model');
const Course = require('./course.model');
const Teacher = require('./teacher.model');

const Student = sequelize.define(
  'student',
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    registrationNumber: {
      type: DataTypes.STRING(50),
      allowNull: false,
      validate: {
        notEmpty: true,
      },
    },
    firstName: {
      type: DataTypes.STRING(50),
      allowNull: false,
      validate: {
        notEmpty: true,
      },
    },
    lastName: {
      type: DataTypes.STRING(50),
      allowNull: false,
      validate: {
        notEmpty: true,
      },
    },
    email: {
      type: DataTypes.STRING(100),
      validate: {
        isEmail: true,
      },
    },
    dateOfBirth: {
      type: DataTypes.DATEONLY,
    },
    gender: {
      type: DataTypes.ENUM('Male', 'Female', 'Other'),
    },
    contactNumber: {
      type: DataTypes.STRING(20),
    },
    address: {
      type: DataTypes.TEXT,
    },
    fingerprint: {
      type: DataTypes.TEXT,
      comment: 'Encrypted fingerprint data of the student',
    },
    active: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
    },
    instituteId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Institute,
        key: 'id',
      },
    },
    departmentId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Department,
        key: 'id',
      },
    },
    sectionId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Section,
        key: 'id',
      },
    },
    addedBy: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Teacher,
        key: 'id',
      },
    },
  },
  {
    timestamps: true,
    indexes: [
      {
        unique: true,
        fields: ['registrationNumber', 'instituteId'],
        name: 'unique_reg_number_per_institute',
      },
    ],
  }
);

// Define associations
Student.belongsTo(Institute, { foreignKey: 'instituteId', as: 'institute' });
Institute.hasMany(Student, { foreignKey: 'instituteId', as: 'students' });

Student.belongsTo(Department, { foreignKey: 'departmentId', as: 'department' });
Department.hasMany(Student, { foreignKey: 'departmentId', as: 'students' });

Student.belongsTo(Section, { foreignKey: 'sectionId', as: 'section' });
Section.hasMany(Student, { foreignKey: 'sectionId', as: 'students' });

Student.belongsTo(Teacher, { foreignKey: 'addedBy', as: 'teacher' });
Teacher.hasMany(Student, { foreignKey: 'addedBy', as: 'students' });

// Many-to-Many relationship between Students and Courses
const StudentCourse = sequelize.define(
  'student_course',
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    studentId: {
      type: DataTypes.INTEGER,
      references: {
        model: Student,
        key: 'id',
      },
    },
    courseId: {
      type: DataTypes.INTEGER,
      references: {
        model: Course,
        key: 'id',
      },
    },
    enrollmentDate: {
      type: DataTypes.DATEONLY,
      defaultValue: DataTypes.NOW,
    },
    status: {
      type: DataTypes.ENUM('Active', 'Dropped', 'Completed'),
      defaultValue: 'Active',
    },
  },
  {
    timestamps: true,
    indexes: [
      {
        unique: true,
        fields: ['studentId', 'courseId'],
        name: 'unique_student_course',
      },
    ],
  }
);

Student.belongsToMany(Course, {
  through: StudentCourse,
  as: 'courses',
  foreignKey: 'studentId',
});

Course.belongsToMany(Student, {
  through: StudentCourse,
  as: 'students',
  foreignKey: 'courseId',
});

module.exports = { Student, StudentCourse };

File: Flutter/lib/data/models/student_model.dart
class Student {
  final int? id;
  final int? serverId;
  final String registrationNumber;
  final String firstName;
  final String lastName;
  final String? email;
  final String? dateOfBirth;
  final String gender;
  final String? contactNumber;
  final String? address;
  final String? fingerprint;
  final bool active;
  final int instituteId;
  final int departmentId;
  final int sectionId;
  final int addedBy;
  final String createdAt;
  final String? updatedAt;
  final bool synced;

  Student({
    this.id,
    this.serverId,
    required this.registrationNumber,
    required this.firstName,
    required this.lastName,
    this.email,
    this.dateOfBirth,
    required this.gender,
    this.contactNumber,
    this.address,
    this.fingerprint,
    required this.active,
    required this.instituteId,
    required this.departmentId,
    required this.sectionId,
    required this.addedBy,
    required this.createdAt,
    this.updatedAt,
    required this.synced,
  });

  // Full name getter
  String get fullName => '$firstName $lastName';

  // Has fingerprint getter
  bool get hasFingerprint => fingerprint != null && fingerprint!.isNotEmpty;

  // Convert from Map (database or API)
  factory Student.fromMap(Map<String, dynamic> map) {
    return Student(
      id: map['id'],
      serverId: map['serverStudentId'],
      registrationNumber: map['registrationNumber'],
      firstName: map['firstName'],
      lastName: map['lastName'],
      email: map['email'],
      dateOfBirth: map['dateOfBirth'],
      gender: map['gender'] ?? 'Other',
      contactNumber: map['contactNumber'],
      address: map['address'],
      fingerprint: map['fingerprint'],
      active: map['active'] == 1,
      instituteId: map['instituteId'],
      departmentId: map['departmentId'],
      sectionId: map['sectionId'],
      addedBy: map['addedBy'],
      createdAt: map['createdAt'],
      updatedAt: map['updatedAt'],
      synced: map['synced'] == 1,
    );
  }

  // Convert to Map for database
  Map<String, dynamic> toMap() {
    return {
      if (id != null) 'id': id,
      if (serverId != null) 'serverStudentId': serverId,
      'registrationNumber': registrationNumber,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'dateOfBirth': dateOfBirth,
      'gender': gender,
      'contactNumber': contactNumber,
      'address': address,
      'fingerprint': fingerprint,
      'active': active ? 1 : 0,
      'instituteId': instituteId,
      'departmentId': departmentId,
      'sectionId': sectionId,
      'addedBy': addedBy,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
      'synced': synced ? 1 : 0,
    };
  }

  // Convert to Map for API
  Map<String, dynamic> toApiMap() {
    return {
      'registrationNumber': registrationNumber,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'dateOfBirth': dateOfBirth,
      'gender': gender,
      'contactNumber': contactNumber,
      'address': address,
      'fingerprint': fingerprint,
      'departmentId': departmentId,
      'sectionId': sectionId,
    };
  }

  // Copy with
  Student copyWith({
    int? id,
    int? serverId,
    String? registrationNumber,
    String? firstName,
    String? lastName,
    String? email,
    String? dateOfBirth,
    String? gender,
    String? contactNumber,
    String? address,
    String? fingerprint,
    bool? active,
    int? instituteId,
    int? departmentId,
    int? sectionId,
    int? addedBy,
    String? createdAt,
    String? updatedAt,
    bool? synced,
  }) {
    return Student(
      id: id ?? this.id,
      serverId: serverId ?? this.serverId,
      registrationNumber: registrationNumber ?? this.registrationNumber,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
      email: email ?? this.email,
      dateOfBirth: dateOfBirth ?? this.dateOfBirth,
      gender: gender ?? this.gender,
      contactNumber: contactNumber ?? this.contactNumber,
      address: address ?? this.address,
      fingerprint: fingerprint ?? this.fingerprint,
      active: active ?? this.active,
      instituteId: instituteId ?? this.instituteId,
      departmentId: departmentId ?? this.departmentId,
      sectionId: sectionId ?? this.sectionId,
      addedBy: addedBy ?? this.addedBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      synced: synced ?? this.synced,
    );
  }
}

class StudentEnrollment {
  final int? id;
  final int? serverId;
  final int studentId;
  final int courseId;
  final String enrollmentDate;
  final String status;
  final String createdAt;
  final String? updatedAt;
  final bool synced;

  StudentEnrollment({
    this.id,
    this.serverId,
    required this.studentId,
    required this.courseId,
    required this.enrollmentDate,
    required this.status,
    required this.createdAt,
    this.updatedAt,
    required this.synced,
  });

  // Convert from Map
  factory StudentEnrollment.fromMap(Map<String, dynamic> map) {
    return StudentEnrollment(
      id: map['id'],
      serverId: map['serverEnrollmentId'],
      studentId: map['studentId'],
      courseId: map['courseId'],
      enrollmentDate: map['enrollmentDate'],
      status: map['status'] ?? 'Active',
      createdAt: map['createdAt'],
      updatedAt: map['updatedAt'],
      synced: map['synced'] == 1,
    );
  }

  // Convert to Map
  Map<String, dynamic> toMap() {
    return {
      if (id != null) 'id': id,
      if (serverId != null) 'serverEnrollmentId': serverId,
      'studentId': studentId,
      'courseId': courseId,
      'enrollmentDate': enrollmentDate,
      'status': status,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
      'synced': synced ? 1 : 0,
    };
  }
}